gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckJUnitDependencies.java

```diff
@@ -0,0 +1,117 @@
+
+package com.palantir.baseline.tasks;
+
+import com.google.common.base.Preconditions;
+import com.palantir.baseline.plugins.BaselineTesting;
+import java.util.Optional;
+import java.util.function.Predicate;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.testing.Test;
+
+public class CheckJUnitDependencies extends DefaultTask {
+
+    public CheckJUnitDependencies() {
+        setGroup("Verification");
+        setDescription("Ensures the correct JUnit4/5 dependencies are present, otherwise tests may silently not run");
+    }
+
+    @TaskAction
+    public final void validateDependencies() {
+        getProject().getConvention()
+                .getPlugin(JavaPluginConvention.class)
+                .getSourceSets()
+                .forEach(ss -> {
+                    Optional<Test> maybeTestTask = BaselineTesting.getTestTaskForSourceSet(getProject(), ss);
+                    if (!maybeTestTask.isPresent()) {
+                        return;
+                    }
+                    Test task = maybeTestTask.get();
+
+                    validateSourceSet(ss, task);
+                });
+    }
+
+    private void validateSourceSet(SourceSet ss, Test task) {
+        boolean junitJupiterIsPresent = hasRuntimeDepMatching(ss, BaselineTesting::isJunitJupiter);
+
+        // When doing an incremental migration to JUnit5, a project may have some JUnit4 and some JUnit5 tests at the
+        // same time. It's crucial that they have the vintage engine set up correctly, otherwise tests may silently
+        // not run!
+        if (sourceSetMentionsJUnit4(ss)) {
+            if (BaselineTesting.useJUnitPlatformEnabled(task)) {
+                Preconditions.checkState(junitJupiterIsPresent,
+                        "Some tests still use JUnit4, but Gradle has been set to use JUnit Platform."
+                                + "To ensure your old JUnit4 tests still run, please add the following:\n\n"
+                                + "    runtimeOnly 'org.junit.jupiter:junit-jupiter'\n\n"
+                                + "Otherwise they will silently not run.");
+
+                boolean vintageEngineExists = hasRuntimeDepMatching(ss, BaselineTesting::isVintageEngine);
+                Preconditions.checkState(vintageEngineExists,
+                        "Some tests still use JUnit4, but Gradle has been set to use JUnit Platform."
+                                + "To ensure your old JUnit4 tests still run, please add the following:\n\n"
+                                + "    runtimeOnly 'org.junit.vintage:junit-vintage-engine'\n\n"
+                                + "Otherwise they will silently not run.");
+            } else {
+                Preconditions.checkState(!junitJupiterIsPresent,
```

Comment:
This might be overly aggressive since other dependencies might pull in junit-jupiter